### PR TITLE
Generate docs for GitHub 2

### PR DESCRIPTION
**Pull request by chidozieononiwu**
.NET Doc generation update

### CODE REVIEW

**Comment by weshaggard** [comment]
@Azure/azure-sdk-eng

File: eng/docgeneration/script/docgen.ps1 [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    New-Item -Type dir dll-yaml-output
+    New-Item -Type dir docfx-output
+}
+
+# Include dependent Dlls and XML files
+function IncludeDependencies($SourcesDirectory, $BinariesDirectory) {
+    Copy-Item -Path "$($SourcesDirectory)/tools/LocalNugetFeed/PopImport.1.0.0.nupkg" -Destination $BinariesDirectory
+    Invoke-WebRequest -Uri "https://www.nuget.org/api/v2/package/Microsoft.Bcl.AsyncInterfaces/1.0.0-preview6.19303.8"`
+    -OutFile "Microsoft.Bcl.AsyncInterfaces.zip" | Wait-Process; Expand-Archive -Path "Microsoft.Bcl.AsyncInterfaces.zip" -DestinationPath "$($BinariesDirectory)/microsoft.bcl.asyncinterfaces"
+    Rename-Item -Path "$($BinariesDirectory)/PopImport.1.0.0.nupkg" -NewName "PopImport.zip"
```
Comment (weshaggard) [inline eng/docgeneration/script/docgen.ps1]: We should try to avoid manually downloading and extracting these packages. Instead we should have a dummy project that we restore and if necessary copy the output after restore.

File: eng/docgeneration/script/docgen.ps1 [hunk 2]
```diff
@@ -0,0 +1,61 @@
+﻿# Extract Dlls from nuget package
+# Copy the items to the dll directory
+function ExtractDlls($PkgDirPath, $DllDirPath) {
```
Comment (weshaggard) [inline eng/docgeneration/script/docgen.ps1]: We should try and avoid extracting these and instead restore them via a project.

File: eng/pipelines/docs.yml [hunk 3]
```diff
@@ -2,85 +2,103 @@ trigger:
...
+        workingDirectory: $(Build.BinariesDirectory)
+        displayName: Install Microsoft.DocAsCode.ECMA2Yaml
       - powershell: |
           Invoke-WebRequest -Uri "https://github.com/dotnet/docfx/releases/download/v2.43.2/docfx.zip" `
           -OutFile "docfx.zip" | Wait-Process; Expand-Archive -Path "docfx.zip" -DestinationPath "./docfx/"
         workingDirectory: $(Build.BinariesDirectory)
         displayName: Download and Extract DocFX
-      - task: CmdLine@2
-        displayName: Provision DocFX Directory
+      - powershell: ". '$(Build.SourcesDirectory)/eng/docgeneration/script/docgen.ps1';CreateDocGenFolders"
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: I would do this as inline ps code. No need to source in the script and then call the function especially for something that isn't really sharing any logic or state.

File: eng/pipelines/docs.yml [hunk 4]
```diff
@@ -2,85 +2,103 @@ trigger:
...
+          buildType: specific
+          project: '590cfd2a-581c-4dcb-a12e-6568ce786175'
+          definition: 37
+          buildVersionToDownload: latestFromBranch
+          tags: Scheduled
+          artifactName: packages
+          targetPath: '$(System.ArtifactsDirectory)/packages'
+      - powershell: ". '$(Build.SourcesDirectory)/eng/docgeneration/script/docgen.ps1';ExtractDlls -PkgDirPath $(System.ArtifactsDirectory)/packages -DllDirPath $(Build.BinariesDirectory)/dll-docs/my-api"
+        displayName: Extract dlls from packages and copy then to dlls directory
+      - powershell: nuget install NETStandard.Library -ExcludeVersion
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: Once you have the wrapper project to do the restore you can also add NETStandard.Library and other nuget packages you need to the closure.

File: eng/docgeneration/script/docgen.ps1 (hunk 1 above)
Comment (chidozieononiwu) [reply eng/docgeneration/script/docgen.ps1]: Updated using dummy project

File: eng/docgeneration/script/docgen.ps1 (hunk 2 above)
Comment (chidozieononiwu) [reply eng/docgeneration/script/docgen.ps1]: Restoring does not generate the .dlls for this, except I build the entire repo which I was trying to avoid.

File: eng/pipelines/docs.yml (hunk 3 above)
Comment (chidozieononiwu) [reply eng/pipelines/docs.yml]: Done

File: eng/pipelines/docs.yml (hunk 4 above)
Comment (chidozieononiwu) [reply eng/pipelines/docs.yml]: Done

File: eng/docgeneration/script/docgen.ps1 (hunk 2 above)
Comment (weshaggard) [reply eng/docgeneration/script/docgen.ps1]: Aren't you extracting binaries from the nuget package? If so you should let nuget do that. You can do your restore against a local set of packages. What I want to try and avoid is you duplicating work that nuget is already good at, we shouldn't hard-code knowledge about the target framework folders and such in these scripts because target frameworks will change overtime but nuget will be able to know the relationship between them and select the correct assets.

File: eng/pipelines/docs.yml [hunk 5]
```diff
@@ -7,75 +7,101 @@ jobs:
...
+        workingDirectory: $(Build.BinariesDirectory)
+        displayName: Download and Extract mdoc Zip
       - pwsh: |
           Invoke-WebRequest -MaximumRetryCount 10 -Uri "https://github.com/dotnet/docfx/releases/download/v2.43.2/docfx.zip" `
           -OutFile "docfx.zip" | Wait-Process; Expand-Archive -Path "docfx.zip" -DestinationPath "./docfx/"
         workingDirectory: $(Build.BinariesDirectory)
         displayName: Download and Extract DocFX
+      - pwsh: |
+          dotnet restore "$(Build.SourcesDirectory)\eng\docgeneration\assets\docgen.csproj"
+          dotnet restore "$(Build.SourcesDirectory)\eng\service.proj"
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: Are you restoring eng\service.proj to get the dependencies?

File: eng/pipelines/docs.yml [hunk 6]
```diff
@@ -7,75 +7,101 @@ jobs:
...
-          workingDirectory: $(Build.SourcesDirectory)
+          script: mdoc.exe fx-bootstrap ../dll-docs
+          workingDirectory: $(Build.BinariesDirectory)/mdoc
           failOnStderr: true
-      - pwsh: ls
-        workingDirectory: $(Build.SourcesDirectory)/docfx_project/
       - pwsh: |
-          Copy-Item "$(Build.SourcesDirectory)/eng/docgeneration/docfx.json" -Destination "$(Build.SourcesDirectory)/docfx_project/" -Force
-        displayName: Copy over docfx.json
+          Copy-Item -Path "$(Build.SourcesDirectory)/restoredPackages/microsoft.bcl.asyncinterfaces/1.0.0-preview6.19303.8/ref/netstandard2.1/Microsoft.Bcl.AsyncInterfaces.dll" `
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: We definitely should not make the assumptions that these packages are restored into the restoredPackages folder that is  a bad assumption. Instead we should write a custom target in one of the projects you are restoring that will copy the files.  See what I did https://dev.azure.com/azure-sdk/internal/_git/azure-sdk-build-tools?path=%2Ftools%2FRestoreToolPackages.proj&version=GBmaster as an example.

File: eng/pipelines/docs.yml (hunk 5 above)
Comment (chidozieononiwu) [reply eng/pipelines/docs.yml]: Yes, I am doing that to get microsoft.bcl.asyncinterfaces and netstandard.library

File: sdk/core/Azure.Core/tests/samples/Sample1_HelloWorld.cs [hunk 7]
```diff
@@ -1,36 +0,0 @@
...
-// Licensed under the MIT License. See License.txt in the project root for
-// license information.
-
-using Azure.Core.Pipeline;
-using NUnit.Framework;
-using System;
-using System.IO;
-using System.Threading.Tasks;
-
-namespace Azure.Core.Samples
```
Comment (pakrym) [inline sdk/core/Azure.Core/tests/samples/Sample1_HelloWorld.cs]: Why was this removed?

Comment (chidozieononiwu) [reply sdk/core/Azure.Core/tests/samples/Sample1_HelloWorld.cs]: That wasn't intended. I will revert that.

File: eng/pipelines/docs.yml [hunk 8]
```diff
@@ -7,75 +7,103 @@ jobs:
...
+        inputs:
+          SourceFolder: '$(Build.ArtifactStagingDirectory)/dll-output'
+          TargetFolder: '$(Build.BinariesDirectory)/dll-docs/my-api'
+          contents: |
+            *.dll
+            *.xml 
+      - task: CopyFiles@2
+        displayName: 'Copy over Api Dependencies'
+        inputs:
+          SourceFolder: '$(Build.SourcesDirectory)/restoredPackages/netstandard.library/2.0.0/build/netstandard2.0/ref'
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: This is making the assumption that the packages are under restoredPackages folder which isn't going to be true longer term as we will be changing that setting. Instead you should copy the files as a post restore step in your docgen.proj file.

Comment (weshaggard) [reply eng/pipelines/docs.yml]: Or perhaps better is you can probably do a publish of that project.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 9]
```diff
@@ -116,7 +116,7 @@ jobs:
           version: "$(DotNetCoreSDKVersion)"
       - script: >-
           dotnet test eng/service.proj --filter TestCategory!=Live --framework $(TestTargetFramework) --logger "trx;LogFileName=$(TestTargetFramework).trx" 
-          /p:ServiceDirectory=${{ parameters.ServiceDirectory }} /p:IncludeSrc=false /p:Configuration=$(BuildConfiguration)
+          /p:ServiceDirectory=${{ parameters.ServiceDirectory }} /p:IncludeSrc=false /p:IncludeSamples=false /p:Configuration=$(BuildConfiguration)
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Is there a reason you need to exclude the samples as part of this work?

File: eng/docgeneration/assets/docgen.csproj [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp2.1</TargetFrameworks>
+  </PropertyGroup>
+  <PropertyGroup>
+    <Description>This is a dummy project used to restore nuget packages used for doc generation</Description>
+    <AssemblyTitle>Doc generation Packages</AssemblyTitle>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.DocAsCode.ECMA2Yaml" Version="1.0.644-alpha" />
+    <PackageReference Include="PopImport" Version="1.0.0" />
```
Comment (weshaggard) [inline eng/docgeneration/assets/docgen.csproj]: Looks like you are still checking in the popimport package as part of this PR we should publish it and not have it committed into the repo.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 9 above)
Comment (chidozieononiwu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: We don't need the samples in the docs.

Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: This is only the test run the docs shouldn't be getting generated from test command line.

File: eng/pipelines/docs.yml [hunk 11]
```diff
@@ -18,8 +18,8 @@ jobs:
         workingDirectory: $(Build.BinariesDirectory)
         displayName: Download and Extract DocFX
       - pwsh: |
-          dotnet restore "$(Build.SourcesDirectory)\eng\docgeneration\assets\docgen.csproj"
-          dotnet restore "$(Build.SourcesDirectory)\eng\service.proj"
+          dotnet restore "$(Build.SourcesDirectory)\eng\docgeneration\assets\docgen.csproj" --packages $(Build.BinariesDirectory)/packages
+          dotnet restore "$(Build.SourcesDirectory)\eng\service.proj" --packages $(Build.BinariesDirectory)/packages
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: While this might help it wasn't the direction I was expecting you to go. Have you tried doing a dotnet publish to get the all the restored binaries into one directory instead of manually copying them? That should work for the libraries you might still need to do some manual copying of the tools given they are in a non-standard layout, but I still believe it is better to do that copying in the docgen.csproj instead of in this yml file.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 9 above)
Comment (chidozieononiwu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: I had to add this due to changes I made in the service.proj. I added a property for Including/excluding samples so I can exclude samples in the docs. I adding that property here to ensure samples are not included or else they will break CI on the NetFramework test

File: eng/docgeneration/assets/docgen.csproj [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    <Description>This is a dummy project used to restore nuget packages used for doc generation</Description>
+    <AssemblyTitle>Doc generation Packages</AssemblyTitle>
+  </PropertyGroup>
+  <PropertyGroup>
+    <ECMA2YmlVersion>1.0.644-alpha</ECMA2YmlVersion>
+    <PopImportVersion>1.0.0</PopImportVersion>
+    <BclAsyncInterfacesVersion>1.0.0-preview8.19405.3</BclAsyncInterfacesVersion>
+    <NETStandardVersion>2.0.2</NETStandardVersion>
+
+    <NuGetPackageRoot>..\..\..\docgenpackages</NuGetPackageRoot>
```
Comment (weshaggard) [inline eng/docgeneration/assets/docgen.csproj]: I suggest you pass in this path as a property to your restore command.

File: eng/pipelines/docs.yml [hunk 13]
```diff
@@ -7,75 +7,103 @@ jobs:
...
+        displayName: Build Packages
+      - task: CopyFiles@2
+        displayName: 'Copy over Dlls'
+        inputs:
+          SourceFolder: '$(Build.ArtifactStagingDirectory)/dll-output'
+          TargetFolder: '$(Build.BinariesDirectory)/dll-docs/my-api'
+          contents: |
+            *.dll
+            *.xml 
+      - task: CopyFiles@2
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: Is this copy necessary? Assuming you pass in the path I wouldn't expect it to be.

File: eng/pipelines/docs.yml [hunk 14]
```diff
@@ -7,75 +7,103 @@ jobs:
...
+          -OutFile "mdoc.zip" | Wait-Process; Expand-Archive -Path "mdoc.zip" -DestinationPath "./mdoc/"
+        workingDirectory: $(Build.BinariesDirectory)
+        displayName: Download and Extract mdoc Zip
       - pwsh: |
           Invoke-WebRequest -MaximumRetryCount 10 -Uri "https://github.com/dotnet/docfx/releases/download/v2.43.2/docfx.zip" `
           -OutFile "docfx.zip" | Wait-Process; Expand-Archive -Path "docfx.zip" -DestinationPath "./docfx/"
         workingDirectory: $(Build.BinariesDirectory)
         displayName: Download and Extract DocFX
+      - pwsh: |
+          dotnet restore "$(Build.SourcesDirectory)/eng/docgeneration/assets/docgen.csproj" --packages docgenpackages
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: I don't think you want to be passing in the `--packages` property to these restore commands any longer.

File: eng/pipelines/docs.yml [hunk 15]
```diff
@@ -7,75 +7,103 @@ jobs:
...
+          New-Item -Type dir dll-docs/dependencies/my-api
+          New-Item -Type dir dll-xml-output
+          New-Item -Type dir dll-yaml-output
+          New-Item -Type dir docfx-output
+        displayName: Create Directories for Doc Generation
+        workingDirectory: $(Build.BinariesDirectory)
+      - task: DotNetCoreInstaller@2
+        displayName: "Use .NET Core sdk 3.0.100-preview5-011568"
+        inputs:
+          version: 3.0.100-preview5-011568
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: We should share this with the globals.yml files to ensure we don't break because of the toolset differences. This has just moved to preview7 for example.

File: eng/pipelines/docs.yml (hunk 13 above)
Comment (chidozieononiwu) [reply eng/pipelines/docs.yml]: Removed

File: eng/pipelines/docs.yml (hunk 14 above)
Comment (chidozieononiwu) [reply eng/pipelines/docs.yml]: Removed

File: eng/pipelines/docs.yml [hunk 16]
```diff
@@ -3,79 +3,98 @@ trigger:
...
       - pwsh: |
           Invoke-WebRequest -MaximumRetryCount 10 -Uri "https://github.com/dotnet/docfx/releases/download/v2.43.2/docfx.zip" `
           -OutFile "docfx.zip" | Wait-Process; Expand-Archive -Path "docfx.zip" -DestinationPath "./docfx/"
         workingDirectory: $(Build.BinariesDirectory)
         displayName: Download and Extract DocFX
+      - pwsh: |
+          dotnet restore "$(Build.SourcesDirectory)/eng/docgeneration/assets/docgen.csproj" /p:BuildBinariesDirectory=$(Build.BinariesDirectory)
+        displayName: Restore Nuget Packages Needed
+      - pwsh: |
+          New-Item -Type dir dll-docs/my-api
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: nit: I'd find this a little easier to read if it was `mkdir dll-docs/my-api`

File: eng/pipelines/docs.yml [hunk 17]
```diff
@@ -3,79 +3,98 @@ trigger:
...
-          Copy-Item "$(Build.SourcesDirectory)/CONTRIBUTING.md" -Destination "$(Build.SourcesDirectory)/docfx_project/articles/"
-          Copy-Item "$(Build.SourcesDirectory)/packages.md" -Destination "$(Build.SourcesDirectory)/docfx_project/articles/"
-          Copy-Item "$(Build.SourcesDirectory)/eng/docgeneration/docfx.json" -Destination "$(Build.SourcesDirectory)/docfx_project/" -Force
+          Copy-Item "$(Build.BinariesDirectory)/dll-yaml-output/*" -Destination "$(Build.BinariesDirectory)/docfx-output/docfx_project/api/" -Recurse
+          Copy-Item "$(Build.SourcesDirectory)/README.md" -Destination "$(Build.BinariesDirectory)/docfx-output/docfx_project/index.md" -Recurse -Force
+          Copy-Item "$(Build.SourcesDirectory)/README.md" -Destination "$(Build.BinariesDirectory)/docfx-output/docfx_project/api/index.md" -Recurse -Force
+          Copy-Item "$(Build.SourcesDirectory)/eng/docgeneration/assets/docfx.json" -Destination "$(Build.BinariesDirectory)/docfx-output/docfx_project" -Recurse -Force
+          Remove-Item -Path "$(Build.BinariesDirectory)/docfx-output/docfx_project/toc.yml" -Force
```
Comment (weshaggard) [inline eng/pipelines/docs.yml]: Is there something else that updates toc.yml? If not why do we need to keep overwriting it? Also is there a reason we don't just check this file in and copy the entire file instead of writing the same content each time?

Comment (chidozieononiwu) [reply eng/pipelines/docs.yml]: The default toc.yml theat DocFx will generate is not what we want hence the overwrite. I was actually trying to avoid checking another file into the repo. Do you want me to check it in?

Comment (weshaggard) [reply eng/pipelines/docs.yml]: Given it is static data I would check it in as it will be easier to understand. Writing to the file like this makes me think that we dynamically generating this and it just introduces more points of potential failure.

**Review by weshaggard — APPROVED** [review]
one last comment both otherwise looks good.

**Review by weshaggard — APPROVED** [review]
